### ISSUE TITLE
Removing the assignment in SimplifyDefUse causes the header to be invalid

### ISSUE THREAD

**Issue by kostasyrmia** [p4lang/p4c#2890]
Here is a piece of code from the gauntlet_hdr_assign_1-bmv2.p4 program:

    control ingress(inout Headers h, inout Meta m, inout standard_metadata_t sm) {
        H x;

        apply {
            x = h.h;
            x.a = 16w2;
            h.eth_hdr.eth_type = x.a;
        }
    }

Here is the output after simplifyDefUse:

    @name("x") H x_0;
    apply
    {
        x_0.setInvalid();
        x_0.a = 16w2;
        h.eth_hdr.eth_type = x_0.a;
    }
 
The assignment `x = h.h` is removed and that causes the header x to always be invalid in this control.

**Comment by fruffy** [comment]
Note that localCopyPropagation later "fixes" the unstable intermediate state. 
```p4
@name("ingress.x") H x_0;
apply {
    x_0.setInvalid();
    x_0.a = 16w2;
    h.eth_hdr.eth_type = 16w2;
}
```

I am not sure about the logic here but LocalCopyPropagation does not consider the validity of headers when propagating values. DefUse seems to operate under that assumption which allows it to remove the statement after the header has been set invalid. 
```p4
@name("x") H x_0;
apply {
    x_0.setInvalid();
    x_0 = h.h;
    x_0.a = 16w2;
    h.eth_hdr.eth_type = x_0.a;
}
```
If you try this with a program that does not have the `x_0.a = 16w2;` statement, DefUse will not remove the assignment as it required for `h.eth_hdr.eth_type = x_0.a` to be correct. 

This is all under the assumption that CopyProp is called, however. The program remains in this unstable state until the MidEnd, where CopyPropagation fixes the issue. If there is a back end that considers assignments to invalid headers no-ops and does not use CopyPropagation this could lead to a bug?

**Comment by mbudiu-vmw** [comment]
Yes, this is an interesting case. Apparently the assignment statement `x_0 = h.h` is useless, since only the value in `x_0.a` is used afterwards. However, the value cannot be written into `x_0.a` unless the previous assignment has been executed. So in some sense all assignments to a header also read the valid flag: if the flag is not set, the assignments may not do what they are expected to do.
